crates/core/tedge_api/src/entity_store.rs

```diff
@@ -676,6 +672,164 @@ impl EntityStore {
+            .iter()
+            .filter_map(|topic_id| self.entities.get_key_value(topic_id))
+            .map(|(k, v)| (k, v.metadata()))
+            .collect()
+    }
+
+    /// Returns information for an entity under a given device/service id.
+    pub fn get_by_external_id(&self, external_id: &EntityExternalId) -> Option<&EntityMetadata> {
+        let topic_id = self.entity_id_index.get(external_id)?;
+        self.get(topic_id)
+    }
+
+    /// Insert a new entity
+    ///
+    /// Return Inserted if the entity is new
+    /// Return Updated if the entity was previously registered and has been updated by this call
+    /// Return Unchanged if the entity not affected by this call
+    pub fn insert(
+        &mut self,
+        external_id: EntityExternalId,
+        topic_id: EntityTopicId,
+        entity_metadata: EntityMetadata,
+    ) -> InsertOutcome {
+        let maybe_parent = entity_metadata.parent.clone();
+        let previous = self.entities.entry(topic_id.clone());
+        let outcome = match previous {
+            Entry::Occupied(mut occupied) => {
+                // if there is no change, no entities were affected
+                let existing_entity = occupied.get().metadata.clone();
+                let existing_children = occupied.get().children.clone();
+
+                let mut merged_other = existing_entity.other.clone();
+                merged_other.extend(entity_metadata.other.clone());
+                let merged_entity = EntityMetadata {
+                    twin_data: existing_entity.twin_data.clone(),
+                    other: merged_other,
+                    ..entity_metadata
+                };
+
+                if existing_entity == merged_entity {
+                    InsertOutcome::Unchanged
+                } else {
+                    let updated_entity = EntityNode {
+                        metadata: merged_entity,
+                        children: existing_children,
+                    };
+                    occupied.insert(updated_entity);
+                    InsertOutcome::Updated
+                }
+            }
+            Entry::Vacant(vacant) => {
+                vacant.insert(EntityNode::new(entity_metadata));
+                self.entity_id_index.insert(external_id, topic_id.clone());
+                InsertOutcome::Inserted
+            }
+        };
+
+        if let Some(parent) = maybe_parent {
+            if let Some(parent_entry) = self.entities.get_mut(&parent) {
+                parent_entry.children.insert(topic_id);
+            }
+        }
+
+        debug!("Updated entity map: {:?}", self.entities);
+        debug!("Updated external id map: {:?}", self.entity_id_index);
+        outcome
+    }
+
+    /// Recursively remove an entity, its child devices and services
+    ///
+    /// Populate the given vector with the topic identifiers of the removed entities
+    pub fn remove(&mut self, topic_id: &EntityTopicId, removed_entities: &mut Vec<EntityTopicId>) {
```

Comment:
Minor: I'd used this signature for an internal function that does the recursion, but exposed a different signature for the external API that just returns a `Vec<EntityTopicId>` instead of accepting the `&mut Vec<EntityTopicId>` as the input. But, since `EntityStore` is an internal API, it doesn't matter that much.

Reply:
What you describe is done by `EntityStore::deregister_entity()`

Reply:
Removing `pub`: https://github.com/thin-edge/thin-edge.io/pull/3237/commits/0f4c40132a64a44522bd454a6f74e86a153348c3